### DISCUSSION TITLE
Pyinstaller Console Deltas

### DISCUSSION THREAD

**Discussion by DRed9X** [pyinstaller/pyinstaller#7815] [PyInstaller] [answered]
Hey Community! First off thanks for Pyinstaller. It's great. 

Question:
What happened to Pyinstaller after v5.5.0 in reference to console=False?

Explanation:
I've been using Pyinstaller to compile apps, but have recently noticed a change in behavior between versions in reference to the "console=False" value. It's nice to not have a hanging command prompt window, but console=True seems to be required after v5.5.0 for any standard output to work. I point standard output to a file and then any line in that file gets inserted into my Tkinter Text Widget.

**With no code changes to my python app this works only on v5.5.0 and earlier.** Updating Pyinstaller to a higher version causes nothing to show up so the behavior is narrowed down to some change in Pyinstaller.

Can anyone explain why this is? I searched the discussion, but the posts close to this scenario seem a little different. Understanding would be greatly appreciated as I would love to use the most up to date version. Excited for 6.0!

Environment:
Windows 11 Pro 22H2, Windows 10 Pro 21H2
Python 3.9.6, Python 3.11.4

**Comment by bwoodsend** [comment]
In PyInstaller 5.7, we stopped replacing `sys.stdout` (which is otherwise `None`) with a placeholder which matches how `pythonw.exe` behaves. It's not clear to me what you mean by *any standard output to work* or *point standard output to a file*. Something like `sys.stdout.write("hello")` used to be allowed (even if it did nothing) but now it will break because `sys.stdout` is `None`. Using `print()` is always safe. If you want to interact with `sys.stdout` directly, wrap it in an `if sys.stdout is not None:` guard.

**Reply by DRed9X** [reply]
@bwoodsend
Thanks for the reply! An example of what I mean would be:

1.  Making a directory using the tempfile module.
2. Creating a file in that directory.
3. Setting sys.stdout to that file. Example:
    sys.stdout = open(filename, "a")
4. Print statements automatically go to standard output which go to the opened file.
5. Reading the lines in that file and sending it to my Tkinter Text Widget.
6. Pyinstaller Spec File: console=False
    This hides the hanging console window.

Hopefully my explanation is understood. I've definitely been using print(). This works with Pyinstaller 5.5.0, but not newer versions. In newer versions I have to have console=True which produces a hanging command prompt window.

#6441 Mentions using the Win32 API, but that still creates a hanging command prompt window that can be interacted with by a user. The most that can be done in this manner is minimizing the command prompt window using the Win32 API.

Your insight is appreciated!

**Reply by rokm** [reply]
Hmmm, if you are overriding `sys.stdout` with custom file-based stream, then PyInstaller >= 5.7 not setting it to dummy stream anymore should not make a difference.

The only way it would make a difference is if you are expecting output to `sys.stdout` from some 3rd party package, and that package is smart enough to disable the output on seeing that `sys.stdout` is `None` (assuming that it checks that during its import, and you override `sys.stdout` with your file-based stream later on).

Can you create a minimal self-contained example that we can use to reproduce this issue?

**Reply by DRed9X** [reply]
@rokm
Thanks for the reply! Here is an example below. self.info_text represents a Tkinter Text Widget and threaded_func is named that way to get across that this would run in it's own thread (called by another func that puts it in it's own thread) in case that matters.

```
def threaded_func(self):
    my_dir = tempfile.mkdtemp()
    my_file = os.path.join(my_dir, ".buf")
    f = open(my_file, "w+") # Creates the file
    while True:
        time.sleep(0.1)
        sys.stdout.flush()  # Force write
        with open(my_file, "r") as buf:
            read = buf.readlines()
            if read:  # Do not write if empty
                for line in read:
                    self.add_info(line)
        with open(my_file, "w"):
            pass  # Clear file
        sys.stdout = open(my_file, "a")

def add_info(self, info):
    self.info_text.config(state="normal")
    now = datetime.datetime.now()
    time = now.strftime("%H:%M:%S")
    info = ">" + f"({time}) " + str(info)
    self.info_text.insert(END, info)
    self.info_text.see(END)
```

Let me know if you need more, a better example, or if you need me to provide compiled examples. I created a simple example Tkinter app with a single button that just uses `print()` and the behavior is consistent with what I've been seeing.

**Reply by rokm** [reply, marked as answer]
I was hoping for the full example, because that would make it clearer where `sys.stdout` gets assigned for the first time.

The above snippet implies that this happens in `threaded_func`, at the end of the while loop; if that is indeed the case, then at the very first iteration, `sys.stdout.flush()` probably raises an exception due to `sys.stdout` being `None` - and with this being a separate thread, the exception does not crash the application, but it does stop the thread from doing any further processing. If you had console, you would probably see that error printed - but if you enable console, `sys.stdout` is not `None` anymore, and the problem goes away...

Can you try adding an `if sys.stdout is not None:` guard to the `sys.stdout.flush()` call? Does that fix the problem?

**Reply by DRed9X** [reply]
@rokm @bwoodsend
Thank you both so much. Adding the guard did in fact work. @bwoodsend I didn't fully understand your explanation at first (apologies) but now I do with the additional details provided by @rokm . I also understand how Pyinstaller works a little better now.

I did in fact have sys.stdout.flush() wrapped in a try/except in the original code, but to your point I never saw the exception. I even successfully compiled an app using the latest Pyinstaller version. Simply setting sys.stdout earlier in the process would also work. I can see this now with my increased understanding.

Again thank you both so much for your time and explanations on Pyinstaller. I look forward to v6.0. Especially  #7713.
